src/lib/AccountFinder.ts

```diff
@@ -0,0 +1,58 @@
+import { AccountInfo } from '@/lib/AccountInfo';
+import { NetworkClient } from '@nimiq/network-client';
+import { WalletStore } from '@/lib/WalletStore';
+
+export default class AccountFinder {
+    public static async findAccounts(
+        deriveAccounts: (startIndex: number, count: number) => Promise<Array<{ address: string, keyPath: string }>>,
+        deriveWalletId: (accounts: Array<{ address: string, keyPath: string, balance: number }>) => Promise<string>,
+        defaultLabel: string = 'Standard Account',
+        onIntermediateResult?: (accounts: Array<{ address: string, keyPath: string, balance: number }>) => void,
+    ): Promise<AccountInfo[]> {
+        await NetworkClient.Instance.init(); // initialize NetworkClient if not initialized yet
+        const foundAccounts: Array<{ address: string, keyPath: string, balance: number }> = [];
+        let nextStartIndex = 0;
+        let derivedAccountsPromise = deriveAccounts(nextStartIndex, AccountFinder.MAX_ALLOWED_GAP);
+        nextStartIndex += AccountFinder.MAX_ALLOWED_GAP;
+        while (true) {
+            const derivedAccounts = await derivedAccountsPromise;
+            if (derivedAccounts.length !== AccountFinder.MAX_ALLOWED_GAP) {
+                throw new Error('\'deriveAccounts\' returned wrong number of accounts');
+            }
+
+            // already start deriving next accounts
+            derivedAccountsPromise = deriveAccounts(nextStartIndex, AccountFinder.MAX_ALLOWED_GAP);
+            nextStartIndex += derivedAccounts.length; // by always advancing in blocks of MAX_ALLOWED_GAP it can happen
+            // that we include accounts that are more than MAX_ALLOWED_GAP apart (e.g. one towards the beginning of
+            // the first block, the other one towards the end of the second block). But that's actually not bad.
+
+            // find accounts with a balance > 0
+            // TODO should use transaction receipts combined with balance
```

Comment:
It's a bit more involved than that, because you are potentially facing rate-limiting by peers when requesting `transactionHistory`, thus it would need additional fail-safes, which are outside the scope of this PR.

I created https://github.com/nimiq/accounts/issues/94 to track it.

Reply:
The nano-api currently supports a method `requestTransactionHistory`, ideally we'd however only ask for transaction receipts, which is much cheaper to calculate.
On the other hand of course one doesn't import accounts that often... So for now `requestTransactionHistory` could be ok.

No, I'd say accounts that have transactions OR a balance. An account that has a balance without (incoming) transactions is a rare corner case though. It would be either a genesis account (these are all legacy wallets though for which we don't have to run the account finding anyways) or accounts that mined a block. So we could also drop the ""OR a balance" part to make the call cheaper.